### PR TITLE
Logbook: Localize Message & Style updates

### PR DESCRIPTION
**Pull request by zsarnett**
## Proposed change

image
00e8135a302167be077c6f454006fdf9
image

## Type of change

- [x] New feature (thank you!)
- [x] Code quality improvements to existing code or addition of tests

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.

Core PR is reliant on the frontend PR merging before it can be merged as it is a breaking change: https://github.com/home-assistant/core/pull/40070

### CODE REVIEW

**Comment by zsarnett** [comment]
zsarnett 2 hours ago • 
Couple things:

 Currently, in the Logbook Element we are basing the icon off of the current state of the entity. Not the state of the entity at the time of the logbook entry
To do this, we need the logbook API to return the state that the entity was set to. Right now we return a message that includes the state but in a "was set to On" form.

 We are unable, at least to my testing and knowledge, to specify more than one entity for the logbook. Its either 1 entity or all entities. I would like this card to work with multiple entities. Example: All motion entities.

@bdraco Worked with you recently to help us with the More info logbook. Would you be able to help me with these changes in the backend to better support this card and the logbook as a whole?

@bdraco
 
Member
bdraco (J. Nick Koston) 2 hours ago
Getting the icon is expensive because we have to decode the json. If there can be only one icon in the attributes we can usually avoid the json decode. I think that's the case, but I need to validate that.

Can you provide a bit more detail on the multi entity use case?

@zsarnett
 
Member
Author
zsarnett 2 hours ago • 
We are already decoding the object that is sent to the frontend. (I don't know much about the backend section) but if you have the was turned on message. I don't know how that's built but if you send in the last on as a state property we get the icon in the frontend. I don't know if that's what you mean when you say expensive.

Use case would be this card. Being able to see a log book for all Motion Sensor entities and building a security dashboard of sorts with the card. All Windows. We would be allowing the users to choose which entities to query for instead of just one or all of them

@bdraco bdraco (J. Nick Koston) mentioned this pull request 1 hour ago
Extract the icon and state for logbook state changed events home-assistant/core#40039
 Open
3 of 21 tasks complete 
@bdraco
 
Member
bdraco (J. Nick Koston) 1 hour ago
For the icon, can you try

home-assistant/core#40039

@bdraco
 
Member
bdraco (J. Nick Koston) 1 hour ago
For the icon, can you try

home-assistant/core#40039

I think that will only work if there is a custom icon set. To calculate the default icon on the frontend, I think you need the 'state' as well?

@bdraco
 
Member
bdraco (J. Nick Koston) 1 hour ago
For the icon, can you try
home-assistant/core#40039

I think that will only work if there is a custom icon set. To calculate the default icon on the frontend, I think you need the 'state' as well?

I adjusted the PR to provide state and icon

@bdraco
 
Member
bdraco (J. Nick Koston) 1 hour ago
We have this giant block to create the message home-assistant/core@aa07171/homeassistant/components/logbook/__init__.py#L551. Ideally we send state from the backend and no message and the frontend will generate the message (localized) instead. This would get us a smaller response.

I'd also like to drop "domain" as it appears to be unused by the frontend and means we end up sending the string again since its already in "entity_id"

@bdraco
 
Member
bdraco (J. Nick Koston) 1 hour ago
I dont think we need to send Icon or domain. We can get those from Hass in the frontend

I can drop icon but I think its needed to fully match what the frontend does:
https://github.com/home-assistant/frontend/blob/dev/src/common/entity/state_icon.ts#L22

@zsarnett
 
Member
Author
zsarnett 1 hour ago
Actually, Icon would be helpful actually. Re reading the code

@zsarnett
 
Member
Author
zsarnett 1 hour ago
We don't need domain. We currently use domain but we don't need to

@bdraco
 
Member
bdraco (J. Nick Koston) 1 hour ago • 
What do you think about doing this part on the frontend?
home-assistant/core@aa07171/homeassistant/components/logbook/__init__.py#L551

Edit: even if the first step is just porting the code and it isn't localized, it would be an improvement as the backend wouldn't have to send those text blobs for each state.

@zsarnett
 
Member
Author
zsarnett 1 hour ago
I much prefer us to do that in the Frontend as we can run translations (I actually don't know if that done in the backend or not)

@bramkragten (who is probably in bed) would be a better fit to say whether we should or not.

I personally think it should be in the frontend though

**Comment by zsarnett** [comment]
I have pulled out the changes for the logbook into here. Original Discussion was here: https://github.com/home-assistant/frontend/pull/6976

**Comment by zsarnett** [comment]
I have started the logbook message in the frontend but it still needs localized

**Comment by zsarnett** [comment]
- [ ] Localize State

**Comment by bdraco** [comment]
Screen Shot 2020-09-14 at 10 03 50 PM


It looks like we are missing icons for start/stop/restart events in the `homeassistant` domain

**Comment by bdraco** [comment]
```
{
	"when": "2020-09-15T02:59:41+00:00",
	"name": "Home Assistant",
	"message": "stopped",
	"domain": "homeassistant"
}, {
	"when": "2020-09-15T03:00:05+00:00",
	"name": "Masterbed TV Cabinet 32 Restart",
	"state": "off",
	"entity_id": "switch.masterbed_tv_cabinet_32_restart",
	"icon": "mdi:restart"
}, {
	"when": "2020-09-15T03:00:05+00:00",
	"name": "Gate Trigger 32 Relay",
	"state": "off",
	"entity_id": "binary_sensor.gate_trigger_32_relay"
}
```

**Comment by zsarnett** [comment]
@bdraco This should be fixed in the latest commit

**Comment by bdraco** [comment]
Looks good.

We did loose the exact time on the more info pop up.  Was that one purpose?

**Comment by zsarnett** [comment]
Yes that was on purpose

File: src/common/entity/domain_icon.ts [hunk 1]
```diff
@@ -71,13 +35,122 @@ export const domainIcon = (domain: string, state?: string): string => {
...
+          return is_off ? "hass:crop-portrait" : "hass:vibrate";
+        case "window":
+          return is_off ? "hass:window-closed" : "hass:window-open";
+        default:
+          return is_off ? "hass:radiobox-blank" : "hass:checkbox-marked-circle";
+      }
+    }
+
+    case "cover": {
+      const open = compareState !== "closed";
```
Comment (bramkragten) [inline src/common/entity/domain_icon.ts:89]: Why did you move all domain-specific icon code in here instead of importing the function?

File: src/data/logbook.ts [hunk 2]
```diff
@@ -72,6 +106,166 @@ const getLogbookDataFromServer = async (
...
   DATA_CACHE[`${startDate}${endDate}`] = {};
 };
+
+export const getLogbookMessage = (
+  hass: HomeAssistant,
+  state: string,
+  stateObj: HassEntity,
+  domain: string
+): string => {
+  const localizePath = "ui.components.logbook.messages";
```
Comment (bramkragten) [inline src/data/logbook.ts:119]: Move it out of the function.

File: src/data/logbook.ts [hunk 3]
```diff
@@ -72,6 +106,166 @@ const getLogbookDataFromServer = async (
...
+            return hass.localize(`${localizePath}.was_disconnected`);
+          }
+          break;
+
+        case "door":
+        case "garage_door":
+        case "opening":
+        case "window":
+          if (isOn) {
+            return hass.localize(`${localizePath}.was_opened`);
```
Comment (bramkragten) [inline src/data/logbook.ts:162]: Can we make the localize names:
`ui.components.logbook.messages.${domain}.${device_class || "_"}.${state}`

That's what we also use for the state translations, and requires a lot less code.

File: src/panels/logbook/ha-logbook.ts [hunk 4]
```diff
@@ -123,58 +128,78 @@ class HaLogbook extends LitElement {
...
+                  : item.context_event_type === "call_service"
+                  ? // Service Call
+                    ` by service
                   ${item.context_domain}.${item.context_service}`
-                : item.context_entity_id === item.entity_id
-                ? // HomeKit or something that self references
-                  ` by
+                  : item.context_entity_id === item.entity_id
+                  ? // HomeKit or something that self references
+                    ` by
```
Comment (bramkragten) [inline src/panels/logbook/ha-logbook.ts:176]: Localize

File: src/panels/logbook/ha-logbook.ts [hunk 5]
```diff
@@ -123,58 +128,78 @@ class HaLogbook extends LitElement {
...
-                  html` by
-                    <a
-                      href="#"
-                      @click=${this._entityClicked}
-                      .entityId=${item.context_entity_id}
-                      class="name"
-                      >${item.context_entity_id_name}</a
-                    >`}
+                  : // Another entity such as an automation or script
+                    html` by
```
Comment (bramkragten) [inline src/panels/logbook/ha-logbook.ts:183]: Localize

File: src/panels/logbook/ha-logbook.ts [hunk 6]
```diff
@@ -123,58 +128,78 @@ class HaLogbook extends LitElement {
...
-                  ` by service
+                  : ""}
+                ${item.message}
+                ${item_username
+                  ? ` by ${item_username}`
+                  : !item.context_event_type
+                  ? ""
+                  : item.context_event_type === "call_service"
+                  ? // Service Call
+                    ` by service
```
Comment (bramkragten) [inline src/panels/logbook/ha-logbook.ts:172]: Localize

File: src/panels/logbook/ha-logbook.ts [hunk 7]
```diff
@@ -123,58 +128,78 @@ class HaLogbook extends LitElement {
...
-                ? ` by ${item_username}`
-                : !item.context_event_type
-                ? ""
-                : item.context_event_type === "call_service"
-                ? // Service Call
-                  ` by service
+                  : ""}
+                ${item.message}
+                ${item_username
+                  ? ` by ${item_username}`
```
Comment (bramkragten) [inline src/panels/logbook/ha-logbook.ts:167]: Localize

File: src/panels/logbook/ha-logbook.ts [hunk 8]
```diff
@@ -123,58 +128,78 @@ class HaLogbook extends LitElement {
...
-        <div class="entry">
-          <div class="time">
-            ${formatTimeWithSeconds(new Date(item.when), this.hass.language)}
-          </div>
+        <div
+          class="entry ${classMap({ "no-entity": !item.entity_id })}"
+          .entityId=${item.entity_id}
+          @click=${this._entityClicked}
+        >
+          ${!this.narrow
```
[the remark this answers is not in the document]
Comment (zsarnett) [reply src/panels/logbook/ha-logbook.ts:136]: hmm I disagree. I was thinking about adding it to the narrow version as well

[the remark this answers is not in the document]
Comment (zsarnett) [reply src/panels/logbook/ha-logbook.ts:136]: Narrow being: Relative_Time - Absolute_Time

Reason being that some people want to know when but I don't want to calculate how many hours/mins ago that was

File: src/common/entity/domain_icon.ts (hunk 1 above)
Comment (zsarnett) [reply src/common/entity/domain_icon.ts:89]: We need to default after wards. If the conditions aren't met we run the last lines of the domain icon

File: src/panels/logbook/ha-logbook.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (bramkragten) [reply src/panels/logbook/ha-logbook.ts:136]: Put the absolute time in a tooltip?

File: src/common/entity/domain_icon.ts (hunk 1 above)
Comment (bramkragten) [reply src/common/entity/domain_icon.ts:89]: Wouldnt this on top be enough? If the domain has a specific icon function let that handle it...
```
  if (domain in domainIcons) {
    return domainIcons[domain](state);	
  }
```

Comment (zsarnett) [reply src/common/entity/domain_icon.ts:89]: Some domains have a specific icon but also non specific. Those are a fallback

File: src/panels/logbook/ha-logbook.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (zsarnett) [reply src/panels/logbook/ha-logbook.ts:136]: image

Tooltip or something like this?

[the remark this answers is not in the document]
Comment (bramkragten) [reply src/panels/logbook/ha-logbook.ts:136]: yeah that could also work, but the logbook lines can be long, and they wrap now, you don't want a secondary line then I guess...

[the remark this answers is not in the document]
Comment (zsarnett) [reply src/panels/logbook/ha-logbook.ts:136]: image

[the remark this answers is not in the document]
Comment (zsarnett) [reply src/panels/logbook/ha-logbook.ts:136]: From that statement of no secondary then you don't want the relative time in the secondary with the tooltip either? You are saying keep it to the left?

[the remark this answers is not in the document]
Comment (zsarnett) [reply src/panels/logbook/ha-logbook.ts:136]: Personally the time to the left makes it look disconnected but idk. I like having both times. Tooltip would be okay. But sometime I want to know the absolute and sometimes I want to know the relative time without having to calculate it

**Comment by zsarnett** [comment]
image

File: src/data/logbook.ts (hunk 3 above)
Comment (zsarnett) [reply src/data/logbook.ts:162]: The state translations are built by core for all the states. Is it even possible to do that here? like in the code there are a lot of if this state do this otherwise this. Doing it your way would require me to do this for all states? no?

Comment (bramkragten) [reply src/data/logbook.ts]: You would move the logic to the translations instead of the code, not more complicated I guess?

**Comment by bramkragten** [comment]
From the screenshot it looks like the entity is no longer clickable? And we should put the icon in the center of the row

**Comment by zsarnett** [comment]
The row is clickable and the icon doesn't look good center for two lines of text 🤔

**Comment by zsarnett** [comment]
image
image

**Comment by bdraco** [comment]
We probably want to skip grouping state changes when we look at individual entities since it will hide things that users are expected to be in the result when the state changes more than once in 15 minutes.    

Maybe a `no_group` param?

I could add support for it once the logbook PRs move forward.  I'm blocked at the moment on adding anything else as it would just be a conflict mess later:
https://github.com/home-assistant/core/pull/40075
https://github.com/home-assistant/core/pull/40387
https://github.com/home-assistant/core/pull/40070

**Comment by zsarnett** [comment]
That sounds reasonable to me.

**Comment by bramkragten** [comment]
Can we move the timing line above the actual log line and have the icon in the middle of the log line? Also I think it is weird the automation is displayed as a link you can click and the entity is not. If I would click the row I would expect to see details about the event I clicked, but I get the more info and current state of the entity. Would a link style not be more logical here?

**Comment by bramkragten** [comment]
Or have the time aligned on the right maybe?

**Comment by zsarnett** [comment]
The link styles just look weird special with two on the same line.

I can put the time on the right. Just made sense to me to have them as the secondary information.

**Comment by zsarnett** [comment]
Actually removing the link was my main reason off the PR

image


This just looks incredibly old and dated

**Comment by bramkragten** [comment]
But in your proposal, you would have a click target (automation) in a click target (entity)? And it would not be clear what a click would do?

**Comment by zsarnett** [comment]
Well in my mind and the reasoning was the main action of the row click was the Entity that was changed. The more info for the most important entity of the row. The Auatomation/script that changed it is supporting information and is displayed as such. Clicking the link more info to the supporting and clicking the row more info to the main entity. 

Is there no better solution than Anchor links 🤮 ... I guess not

File: src/data/logbook.ts (hunk 3 above)
Comment (zsarnett) [reply src/data/logbook.ts]: How do I do the logic in Json?

**Comment by bdraco** [comment]
Once this is good to go, I can start working on making the triggers localizable.

**Comment by bramkragten** [comment]
Does this need a backend change or is this a bug? Am getting an error:
```
2020-09-28 13:20:08 ERROR (SyncWorker_27) [homeassistant.components.recorder.util] Error executing query: unsupported operand type(s) for |: 'NoneType' and 'BinaryExpression'
2020-09-28 13:20:08 ERROR (MainThread) [aiohttp.server] Error handling request
Traceback (most recent call last):
  File "/Users/bram/Documents/projects/home-assistant/core/venv/lib/python3.7/site-packages/aiohttp/web_protocol.py", line 418, in start
    resp = await task
  File "/Users/bram/Documents/projects/home-assistant/core/venv/lib/python3.7/site-packages/aiohttp/web_app.py", line 458, in _handle
    resp = await handler(request)
  File "/Users/bram/Documents/projects/home-assistant/core/venv/lib/python3.7/site-packages/aiohttp/web_middlewares.py", line 119, in impl
    return await handler(request)
  File "/Users/bram/Documents/projects/home-assistant/core/homeassistant/components/http/request_context.py", line 18, in request_context_middleware
    return await handler(request)
  File "/Users/bram/Documents/projects/home-assistant/core/homeassistant/components/http/ban.py", line 72, in ban_middleware
    return await handler(request)
  File "/Users/bram/Documents/projects/home-assistant/core/homeassistant/components/http/auth.py", line 127, in auth_middleware
    return await handler(request)
  File "/Users/bram/Documents/projects/home-assistant/core/homeassistant/components/http/view.py", line 129, in handle
    result = await result
  File "/Users/bram/Documents/projects/home-assistant/core/homeassistant/components/logbook/__init__.py", line 254, in get
    return await hass.async_add_executor_job(json_events)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/bram/Documents/projects/home-assistant/core/homeassistant/components/logbook/__init__.py", line 250, in json_events
    entity_matches_only,
  File "/Users/bram/Documents/projects/home-assistant/core/homeassistant/components/logbook/__init__.py", line 477, in _get_events
    filters.entity_filter() | (Events.event_type != EVENT_STATE_CHANGED)
TypeError: unsupported operand type(s) for |: 'NoneType' and 'BinaryExpression'
```

**Comment by bramkragten** [comment]
Still think the icons should be vertically centered

**Comment by zsarnett** [comment]
Works for me on the latest core dev

**Comment by bdraco** [comment]
> Does this need a backend change or is this a bug? Am getting an error:
> 

Does this still happen with https://github.com/home-assistant/core/pull/40387 ?

**Comment by bdraco** [comment]
> > Does this need a backend change or is this a bug? Am getting an error:
> 
> Does this still happen with [home-assistant/core#40387](https://github.com/home-assistant/core/pull/40387) ?

@bramkragten  Fixed here https://github.com/home-assistant/core/pull/40565.     [`b7859a0` (#40565)](https://github.com/home-assistant/core/pull/40565/commits/b7859a0a3124fd1a13e0b7e62d6b56c9ccb9b144)

File: src/panels/logbook/ha-logbook.ts [hunk 9]
```diff
@@ -188,110 +209,131 @@ class HaLogbook extends LitElement {
...
+          align-items: center;
+        }
+
+        .no-entries {
+          text-align: center;
+          color: var(--secondary-text-color);
+        }
+
+        ha-icon {
+          margin: 4px 8px 0 0;
```
Comment (bramkragten) [inline src/panels/logbook/ha-logbook.ts:298]:
```suggestion
          margin-right: 16px;
```

Comment (zsarnett) [reply src/panels/logbook/ha-logbook.ts:298]:
```suggestion
          margin-right: 8px;
```

Why do you think they icon needs to be so far from the text?

Comment (bramkragten) [reply src/panels/logbook/ha-logbook.ts:298]: equal from both sides and material guidelines <https://material.io/components/lists#specs>

**Review by bramkragten — APPROVED** [review]
